File: platform/Apple/Sources/SweetEditorCoreInternal/api/EditorSettings.swift
import Foundation

public enum FoldArrowMode: Int32, Equatable {
    case auto = 0
    case always = 1
    case hidden = 2
}

public enum WrapMode: Int32, Equatable {
    case none = 0
    case charBreak = 1
    case wordBreak = 2
}

public enum CurrentLineRenderMode: Int32, Equatable {
    case background = 0
    case border = 1
    case none = 2
}

public enum AutoIndentMode: Int32, Equatable {
    case none = 0
    case keepIndent = 1
}

protocol EditorSettingsHost: AnyObject {
    func applyEditorSettings(_ settings: EditorSettings)
}

public final class EditorSettings {
    private weak var host: EditorSettingsHost?

    public private(set) var editorTextSize: Float = 14.0
    public private(set) var typeface: String = "Menlo"
    public private(set) var scale: Float = 1.0
    public private(set) var foldArrowMode: FoldArrowMode = .always
    public private(set) var wrapMode: WrapMode = .none
    public private(set) var lineSpacingAdd: Float = 0.0
    public private(set) var lineSpacingMult: Float = 1.0
    public private(set) var contentStartPadding: Float = 0.0
    public private(set) var showSplitLine = true
    public private(set) var currentLineRenderMode: CurrentLineRenderMode = .background
    public private(set) var autoIndentMode: AutoIndentMode = .keepIndent
    public private(set) var readOnly = false
    public private(set) var maxGutterIcons: UInt32 = 0
    public private(set) var decorationScrollRefreshMinIntervalMs: Int64 = 16
    public private(set) var decorationOverscanViewportMultiplier: Float = 1.5

    init(host: EditorSettingsHost?) {
        self.host = host
    }

    func attachHost(_ host: EditorSettingsHost) {
        self.host = host
    }

    public func setEditorTextSize(_ textSize: Float) {
        editorTextSize = textSize
        apply()
    }

    public func setTypeface(_ typeface: String) {
        self.typeface = typeface
        apply()
    }

    public func setScale(_ scale: Float) {
        self.scale = scale
        apply()
    }

    public func setFoldArrowMode(_ mode: FoldArrowMode) {
        foldArrowMode = mode
        apply()
    }

    public func setWrapMode(_ mode: WrapMode) {
        wrapMode = mode
        apply()
    }

    public func setLineSpacing(add: Float, mult: Float) {
        lineSpacingAdd = add
        lineSpacingMult = mult
        apply()
    }

    public func setContentStartPadding(_ padding: Float) {
        contentStartPadding = max(0, padding)
        apply()
    }

    public func setShowSplitLine(_ show: Bool) {
        showSplitLine = show
        apply()
    }

    public func setCurrentLineRenderMode(_ mode: CurrentLineRenderMode) {
        currentLineRenderMode = mode
        apply()
    }

    public func setAutoIndentMode(_ mode: AutoIndentMode) {
        autoIndentMode = mode
        apply()
    }

    public func setReadOnly(_ readOnly: Bool) {
        self.readOnly = readOnly
        apply()
    }

    public func setMaxGutterIcons(_ count: UInt32) {
        maxGutterIcons = count
        apply()
    }

    public func setDecorationScrollRefreshMinIntervalMs(_ intervalMs: Int64) {
        decorationScrollRefreshMinIntervalMs = max(0, intervalMs)
    }

    public func setDecorationOverscanViewportMultiplier(_ multiplier: Float) {
        decorationOverscanViewportMultiplier = max(0, multiplier)
    }

    private func apply() {
        host?.applyEditorSettings(self)
    }
}

extension SweetEditorCore.AutoIndentMode {
    init(_ mode: AutoIndentMode) {
        switch mode {
        case .none:
            self = .none
        case .keepIndent:
            self = .keepIndent
        }
    }
}

extension SweetEditorCore.FoldArrowMode {
    init(_ mode: FoldArrowMode) {
        switch mode {
        case .auto:
            self = .auto
        case .always:
            self = .always
        case .hidden:
            self = .hidden
        }
    }
}

extension SweetEditorCore.WrapMode {
    init(_ mode: WrapMode) {
        switch mode {
        case .none:
            self = .none
        case .charBreak:
            self = .charBreak
        case .wordBreak:
            self = .wordBreak
        }
    }
}

File: platform/Android/sweeteditor/src/main/java/com/qiplat/sweeteditor/EditorSettings.java
package com.qiplat.sweeteditor;

import android.graphics.Typeface;

import androidx.annotation.NonNull;

import com.qiplat.sweeteditor.core.foundation.AutoIndentMode;
import com.qiplat.sweeteditor.core.foundation.CurrentLineRenderMode;
import com.qiplat.sweeteditor.core.foundation.FoldArrowMode;
import com.qiplat.sweeteditor.core.foundation.WrapMode;

/**
 * Centralized configuration for {@link SweetEditor}.
 * <p>
 * Obtain via {@link SweetEditor#getSettings()}. All setters take effect immediately.
 */
public class EditorSettings {

    private final SweetEditor mEditor;

    private float mTextSize = 36f;
    private Typeface mTypeface = Typeface.create(Typeface.MONOSPACE, Typeface.NORMAL);
    private float mScale = 1.0f;
    private FoldArrowMode mFoldArrowMode = FoldArrowMode.ALWAYS;
    private WrapMode mWrapMode = WrapMode.NONE;
    private float mLineSpacingAdd = 0f;
    private float mLineSpacingMult = 1.0f;
    private float mContentStartPadding = 0f;
    private boolean mShowSplitLine = true;
    private boolean mGutterSticky = true;
    private CurrentLineRenderMode mCurrentLineRenderMode = CurrentLineRenderMode.BACKGROUND;
    private AutoIndentMode mAutoIndentMode = AutoIndentMode.NONE;
    private boolean mReadOnly = false;
    private int mMaxGutterIcons = 0;
    private long mDecorationScrollRefreshMinIntervalMs = 16L;
    private float mDecorationOverscanViewportMultiplier = 1.5f;

    EditorSettings(@NonNull SweetEditor editor) {
        mEditor = editor;
    }

    public void setEditorTextSize(float textSize) {
        mTextSize = textSize;
        mEditor.applyTextSize(textSize);
    }

    public float getEditorTextSize() {
        return mTextSize;
    }

    public void setTypeface(@NonNull Typeface typeface) {
        mTypeface = typeface;
        mEditor.applyTypeface(typeface);
    }

    @NonNull
    public Typeface getTypeface() {
        return mTypeface;
    }

    public void setScale(float scale) {
        mScale = scale;
        mEditor.getEditorCore().setScale(scale);
        mEditor.syncPlatformScale(scale);
        mEditor.flush();
    }

    public float getScale() {
        return mScale;
    }

    public void setFoldArrowMode(@NonNull FoldArrowMode mode) {
        mFoldArrowMode = mode;
        mEditor.getEditorCore().setFoldArrowMode(mode.value);
    }

    @NonNull
    public FoldArrowMode getFoldArrowMode() {
        return mFoldArrowMode;
    }

    public void setWrapMode(@NonNull WrapMode mode) {
        mWrapMode = mode;
        mEditor.getEditorCore().setWrapMode(mode.value);
        mEditor.flush();
    }

    @NonNull
    public WrapMode getWrapMode() {
        return mWrapMode;
    }

    public void setLineSpacing(float add, float mult) {
        mLineSpacingAdd = add;
        mLineSpacingMult = mult;
        mEditor.getEditorCore().setLineSpacing(add, mult);
        mEditor.flush();
    }

    public float getLineSpacingAdd() {
        return mLineSpacingAdd;
    }

    public float getLineSpacingMult() {
        return mLineSpacingMult;
    }

    public void setContentStartPadding(float padding) {
        mContentStartPadding = Math.max(0f, padding);
        mEditor.getEditorCore().setContentStartPadding(mContentStartPadding);
        mEditor.flush();
    }

    public float getContentStartPadding() {
        return mContentStartPadding;
    }

    public void setShowSplitLine(boolean show) {
        mShowSplitLine = show;
        mEditor.getEditorCore().setShowSplitLine(show);
        mEditor.flush();
    }

    public boolean isShowSplitLine() {
        return mShowSplitLine;
    }

    public void setGutterSticky(boolean sticky) {
        mGutterSticky = sticky;
        mEditor.getEditorCore().setGutterSticky(sticky);
        mEditor.flush();
    }

    public boolean isGutterSticky() {
        return mGutterSticky;
    }

    public void setCurrentLineRenderMode(@NonNull CurrentLineRenderMode mode) {
        mCurrentLineRenderMode = mode;
        mEditor.getEditorCore().setCurrentLineRenderMode(mode.value);
        mEditor.flush();
    }

    @NonNull
    public CurrentLineRenderMode getCurrentLineRenderMode() {
        return mCurrentLineRenderMode;
    }

    public void setAutoIndentMode(@NonNull AutoIndentMode mode) {
        mAutoIndentMode = mode;
        mEditor.getEditorCore().setAutoIndentMode(mode.value);
    }

    @NonNull
    public AutoIndentMode getAutoIndentMode() {
        return mAutoIndentMode;
    }

    public void setReadOnly(boolean readOnly) {
        mReadOnly = readOnly;
        mEditor.getEditorCore().setReadOnly(readOnly);
    }

    public boolean isReadOnly() {
        return mReadOnly;
    }

    public void setMaxGutterIcons(int count) {
        mMaxGutterIcons = count;
        mEditor.getEditorCore().setMaxGutterIcons(count);
    }

    public int getMaxGutterIcons() {
        return mMaxGutterIcons;
    }

    public void setDecorationScrollRefreshMinIntervalMs(long intervalMs) {
        mDecorationScrollRefreshMinIntervalMs = Math.max(0L, intervalMs);
        mEditor.requestDecorationRefresh();
    }

    public long getDecorationScrollRefreshMinIntervalMs() {
        return mDecorationScrollRefreshMinIntervalMs;
    }

    public void setDecorationOverscanViewportMultiplier(float multiplier) {
        mDecorationOverscanViewportMultiplier = Math.max(0f, multiplier);
        mEditor.requestDecorationRefresh();
    }

    public float getDecorationOverscanViewportMultiplier() {
        return mDecorationOverscanViewportMultiplier;
    }
}
